File: Leetcode/typescript/sol110.ts




/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
    if (!root){
        return true;
    }
    const S:[TreeNode | null, boolean][] = [[root, false]];
    const heightMap = new Map<TreeNode | null, number>();
    while(0 < S.length){
        const [node, visited] = S.pop();
        if(!node){
            continue;
        }

        if (!visited){
            S.push([node, true])//processed
            S.push([node.left, false])
            S.push([node.right, false])
        }else{
            let l_height = heightMap.get(node.left) ?? 0;
            let r_height = heightMap.get(node.right) ?? 0;
            if(Math.abs(l_height - r_height) > 1){
                return false;
            }
            heightMap.set(node, 1 + Math.max(l_height,r_height ))
        }
    }
    return true;
};





function isBalanced(root: TreeNode | null): boolean {
    if (!root){
        return true;
    }

    const checkHeight = (node) => {
        if (!node){
            return 0;
        }
        let leftHeight = checkHeight(node.left)
        if (leftHeight == -1){
            return -1;
        }

        let rightHeight = checkHeight(node.right);
        if (rightHeight == -1){
            return -1
        }

        if(Math.abs(rightHeight - leftHeight) > 1){
            return -1;
        }

        return 1 + Math.max(leftHeight, rightHeight)
    }

    return checkHeight(root) != -1;
}

File: Leetcode/python/sol110.py
# https://leetcode.com/problems/balanced-binary-tree/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def helper(node):
            if not node:
                return 0
            left = helper(node.left)
            if left == -1:
                return -1
            right = helper(node.right)
            if right == -1:
                return -1
            if abs(left-right) > 1:
                return -1

            return 1 + max(left, right)

        return helper(root) != -1


        # if not root:
        #     return True

        # S = [(root, False)]
        # heights = {}
        # while S:
        #     node, visited = S.pop()
        #     if not node:
        #         continue
        #     if not visited:
        #         # if not visited, then add the node, and child to the stack
        #         S.append((node, True))
        #         S.append((node.left, False))
        #         S.append((node.right, False))
        #     else:
        #         l_height = heights.get(node.left, 0)
        #         r_height = heights.get(node.right, 0)
        #         if abs(l_height-r_height) > 1:
        #             return False
        #         heights[node] = 1+max(l_height,r_height)
        # return True

